### PR TITLE
feat: List.SatisfiesM_foldlM

### PR DESCRIPTION
**Pull request by kim-em**

### CODE REVIEW

**Comment by kim-em** [comment]
In conjunction with #1029, and https://github.com/leanprover/lean4/pull/6023, we can do things like:

```
def addRandomEvens (n : Nat) (k : Nat) : IO Nat := do
  let mut r := k
  for _ in List.range n do
    r := r + 2 * (← IO.rand 0 37)
  pure r

theorem addRandomEvens_spec (n k) : SatisfiesM (fun r => r % 2 = k % 2) (addRandomEvens n k) := by
  rw [addRandomEvens]
  simp [List.forIn_yield_eq_foldlM]
  apply List.SatisfiesM_foldlM
  · rfl
  · intros b w a m
    apply SatisfiesM.bind_pre
    simp_all [SatisfiesM_EStateM_eq, EStateM.run]

def addRandomEvens' (n : Nat) (k : Nat) : IO { r : Nat // r % 2 = k % 2 } := do
  satisfying (addRandomEvens_spec n k)
```

File: Batteries/Data/List/Monadic.lean [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import Batteries.Classes.SatisfiesM
+
+/-!
+# Results about monadic operations on `List`, in terms of `SatisfiesM`.
+
+-/
+
+namespace List
+
+theorem SatisfiesM_foldlM [Monad m] [LawfulMonad m] {f : β → α → m β} (h₀ : motive b)
```
Comment (eric-wieser) [inline Batteries/Data/List/Monadic.lean:15]:
```suggestion
theorem satisfiesM_foldlM [Monad m] [LawfulMonad m] {f : β → α → m β} (h₀ : motive b)
```

File: Batteries/Data/List/Monadic.lean [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  have g a am b hb := Classical.indefiniteDescription _ (h₁ a am b hb)
+  clear h₁
+  induction l with
+  | nil => exact SatisfiesM.pure h₀
+  | cons hd tl ih =>
+    simp only [foldrM_cons]
+    apply SatisfiesM.bind_pre
+    let ⟨q, qh⟩ := ih (fun a am b hb => g a (mem_cons_of_mem hd am) b hb)
+    exact ⟨(fun ⟨b, bh⟩ => ⟨b, (g hd (mem_cons_self hd tl) b bh).existsOfSubtype⟩) <$> q,
+      by simpa using qh⟩
```
Comment (eric-wieser) [inline Batteries/Data/List/Monadic.lean:41]: Why the classical dance here?

```suggestion
  induction l with
  | nil => exact SatisfiesM.pure h₀
  | cons hd tl ih =>
    simp only [foldrM_cons]
    apply SatisfiesM.bind_pre
    let ⟨q, qh⟩ := ih (fun a am b hb => h₁ a (mem_cons_of_mem hd am) b hb)
    exact ⟨(fun ⟨b, bh⟩ => ⟨b, h₁ hd (mem_cons_self hd tl) b bh⟩) <$> q,
      by simpa using qh⟩
```

File: Batteries/Data/List/Monadic.lean [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  induction l generalizing b with
+  | nil => exact SatisfiesM.pure h₀
+  | cons hd tl ih =>
+    simp only [foldlM_cons]
+    apply SatisfiesM.bind_pre
+    let ⟨q, qh⟩ := g b h₀ hd (mem_cons_self hd tl)
+    exact ⟨(fun ⟨b, bh⟩ => ⟨b, ih bh (fun b bh a am => g b bh a (mem_cons_of_mem hd am))⟩) <$> q,
+      by simpa using qh⟩
+
+theorem SatisfiesM_foldrM [Monad m] [LawfulMonad m] {f : α → β → m β} (h₀ : motive b)
```
Comment (fgdorais) [inline Batteries/Data/List/Monadic.lean:29]:
```suggestion
theorem satisfiesM_foldrM [Monad m] [LawfulMonad m] {f : α → β → m β} (h₀ : motive b)
```

**Review by fgdorais — APPROVED** [review]

**Pull request merged**